### ISSUE TITLE
Subdivide hdf file into train, validation, and test sets

### ISSUE THREAD

**Issue by Christinele14** [gregory-kyro/HAC-Net#2]
Thank you for your great work. I am studying your work and getting confused of how you separate the HDF file which contains the total data of PDBBind v.2020 into training, validation, and test set. I saw in the tutorial .ipynb file you used core_2016.hdf file as the test set. So, could you explain to me more about this step? Thank you in advance.

**Comment by gregory-kyro** [comment]
Hi Christine, we provide all of the HDF files for which we report results in our preprint <https://arxiv.org/abs/2212.12440>. This includes training, validation and test sets in HDF format for each set of results that are presented. You can find all of the HDF files here: <https://drive.google.com/drive/folders/1yB2voUxwzhrQRh0JXnOD3BzY8ZQrbgUK?usp=sharing> within the folder "train_val_test_files". Please let me know if this addresses your concern. 

If you are instead looking to create different train-validation-test splits than those that we used, I would be happy to send you a script to generate the different sets from the original HDF file containing all of the protein-ligand data from the PDBbind v.2020 database. In this case, we would create a new HDF file for each set by writing selected data from the original HDF with all of the data to a new HDF file. I hope this helps.

**Comment by Christinele14** [comment]
Thank you for your fast response and careful explanation. Yes, I have already looked at the training, validation, and test sets in HDF format in your google drive. However, I am trying to preprocess data step by step to understand more about the work. It will be great if you could share with me the script to generate different sets with selected data from the original HDF which contains all data from the PDBbind .v2020 database. Thank you in advance for your help.

**Comment by gregory-kyro** [comment]
Absolutely! You can generate different sets with selected data from the original HDF file using this code:

    # create empty arrays to store PDB IDs for each set
    train_pdbids = []
    val_pdbids = []
    test_pdbids = []
    
    # do whatever you want here to choose the complexes for each set, fill the arrays above with the PDB IDs
    
    
    # create three HDF files and initialize them to be written into
    with h5py.File("/content/drive/training.hdf", 'w') as g, \
    h5py.File("/content/drive/validation.hdf", 'w') as h, \
    h5py.File("/content/drive/test.hdf", 'w') as i:
    
        # read in HDF file with all of the data
        with h5py.File("/path_to_hdf_with_all_data.hdf", 'r') as f:
    
            # write training data to HDF file
            for pdbid in train_pdbids:
                ds = g.create_dataset(pdbid, data=f[pdbid], compression = 'lzf')
                ds.attrs['affinity'] = f[pdbid].attrs['affinity']
                ds.attr['van_der_waals'] = f[pdbid].attrs['van_der_waals']
    
            # write validation data to HDF file
            for pdbid in val_pdbids:
                ds = h.create_dataset(pdbid, data=f[pdbid], compression = 'lzf')
                ds.attrs['affinity'] = f[pdbid].attrs['affinity']
                ds.attr['van_der_waals'] = f[pdbid].attrs['van_der_waals']

            # write test data to HDF file
            for pdbid in test_pdbids:
                ds = i.create_dataset(pdbid, data=f[pdbid], compression = 'lzf')
                ds.attrs['affinity'] = f[pdbid].attrs['affinity']
                ds.attr['van_der_waals'] = f[pdbid].attrs['van_der_waals']

Please let me know if you have any questions! This code should work, you just have to specify which protein-ligand complexes you want to store in each set by filling the initial arrays with desired PDB IDs. If you want, you can read in the HDF files that we provide and save the PDB IDs from each file to a list, allowing you to easily select the same complexes that we did. I hope this helps.

**Comment by Christinele14** [comment]
Thank you for your sharing. I will examine the original HDF files and use your suggested code to prepare training, validation, and test sets from the HDF file which contains all data of PDBbind v.2020. Once again, thank you for your kind help and instruction.
